src/middlewared/middlewared/plugins/migration.py

```diff
@@ -0,0 +1,52 @@
+import asyncio
+import os
+
+from middlewared.service import Service
+from middlewared.utils import load_modules
+
+
+def load_migrations(middleware):
+    main_sources_dir = os.path.join(os.path.dirname(os.path.realpath(__file__)), os.path.pardir, "migration")
+    sources_dirs = [os.path.join(overlay_dir, "migration") for overlay_dir in middleware.overlay_dirs]
+    sources_dirs.insert(0, main_sources_dir)
+
+    modules = []
+    for sources_dir in sources_dirs:
+        modules.extend(load_modules(sources_dir))
+
+    return sorted(modules, key=lambda module: module.__name__)
+
+
+class MigrationService(Service):
+
+    class Config:
+        private = True
+
+    async def run(self):
+        if await self.middleware.call("keyvalue.get", "fake_migration", False):
+            for module in load_migrations(self.middleware):
+                await self.middleware.call("datastore.insert", "system.migration", {"name": module.__name__})
+
+            await self.middleware.call("keyvalue.set", "fake_migration", False)
```

Comment:
I thought you said something yesterday about filling these during build time, did you change your mind?

Reply:
After you said that we are going to do things differently (hopely the proper way) in 12.0, I've changed my mind.